sbnana/SBNAna/NuMINumuXSec/ICARUSNumuXsec_HistoProducer.C

```diff
@@ -0,0 +1,561 @@
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrackMCSP);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrackCombinedP);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrackLength);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrackChi2Proton);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrackReducedChi2Proton);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrack_Length_vs_Chi2Proton);
+  map_cutName_to_vec_Spectrums[currentCutName].push_back(sTruthProtonMatchedTrack_Length_vs_ReducedChi2Proton);
+  //map_cutName_to_vec_Spectrums[currentCutName].push_back(sNStub);
+
+  if(ISysts.size()>0){
+
+    EnsembleSpectrum *esNeutrinoTruthE = new EnsembleSpectrum(loader, axNeutrinoTruthE, spillCut, cut, systWs);
+    EnsembleSpectrum *esTruthQ2 = new EnsembleSpectrum(loader, axTruthQ2, spillCut, cut, systWs);
+    EnsembleSpectrum *esMuonTruthP = new EnsembleSpectrum(loader, axMuonTruthP, spillCut, cut, systWs);
+    EnsembleSpectrum *esMuonTruthCosineTheta = new EnsembleSpectrum(loader, axMuonTruthCosineTheta, spillCut, cut, systWs);
+    EnsembleSpectrum *esProtonTruthP = new EnsembleSpectrum(loader, axProtonTruthP, spillCut, cut, systWs);
+
+		map_cutName_to_vec_EnsembleSpectrums[currentCutName].push_back(esNeutrinoTruthE);
+		map_cutName_to_vec_EnsembleSpectrums[currentCutName].push_back(esTruthQ2);
+		map_cutName_to_vec_EnsembleSpectrums[currentCutName].push_back(esMuonTruthP);
+		map_cutName_to_vec_EnsembleSpectrums[currentCutName].push_back(esMuonTruthCosineTheta);
+		map_cutName_to_vec_EnsembleSpectrums[currentCutName].push_back(esProtonTruthP);
+  }
+
+  cout << "[HistoProducer::bookTruth] Finished" << endl;
+
+}
+
+void HistoProducer::bookRecoMuon(SpectrumLoader& loader, Cut cut, SpillCut spillCut){
+
+  cout << "[HistoProducer::bookRecoMuon] called" << endl;
+
+  //==== some binnings
+  double xEnergyMin = 0.; // GeV
+  double xEnergyMax = 10.;
+  double dxEnergy = 0.1;
+  const Binning binsEnergy = Binning::Simple( int( (xEnergyMax-xEnergyMin)/dxEnergy ), xEnergyMin, xEnergyMax );
+  double xCosineThetaMin = -1;
+  double xCosineThetaMax = 1.;
+  double dxCosineTheta = 0.05;
+  const Binning binsCosineTheta = Binning::Simple( int( (xCosineThetaMax-xCosineThetaMin)/dxCosineTheta ), xCosineThetaMin, xCosineThetaMax );
+  const Binning binsEResidual = Binning::Simple(600, -3., 3.);
+  const Binning binsEResidualFraction = Binning::Simple(600, -3., 3.);
+  const Binning binsPDG = Binning::Simple(100000, -5000., 5000.);
+  const Binning binsLength = Binning::Simple(2000, 0., 2000.);
+  const Binning binsChi2 = Binning::Simple(200, 0., 200);
+  const Binning binsReducedChi2 = Binning::Simple(100, 0., 10);
+
+  cout << "[HistoProducer::bookRecoMuon] Delcaring HistAxis" << endl;
+
+  //==== HistAxis
+  const HistAxis axMuonRecoP("MuonRecoP", binsEnergy, varMuonRecoP);
```

Comment:
I think if you put the axes in a map, then you could loop over them and make and name the spectra automatically, rather than having to write out all the combinations?

Reply:
One HistAxis can be used to create both 1D and 2D Spectrum (e.g., https://github.com/cjbackhouse/sbnana/blob/610ec9227758751469ff47d1ebd7dd00c2e4f735/sbnana/SBNAna/NuMINumuXSec/ICARUSNumuXsec_HistoProducer.C#L122 and https://github.com/cjbackhouse/sbnana/blob/610ec9227758751469ff47d1ebd7dd00c2e4f735/sbnana/SBNAna/NuMINumuXSec/ICARUSNumuXsec_HistoProducer.C#L125), so I'm currently doing in this way..